### ISSUE TITLE
How to invert image into noise using class diffusers.DDIMInverseScheduler?

### ISSUE THREAD

**Issue by g-jing** [huggingface/diffusers#3267]
I wonder how to use class `diffusers.DDIMInverseScheduler` to invert image into image noise? I did not see an example of how to use it. Thanks

**Comment by sayakpaul** [comment]
We have a pipeline [here](https://huggingface.co/docs/diffusers/api/pipelines/stable_diffusion/pix2pix_zero#based-on-an-input-image) that internally makes use of the inverse scheduler -- it might come as a good reference. But I agree that having a standalone example might be helpful. 

@patrickvonplaten do you think you could help add one?

**Comment by g-jing** [comment]
@sayakpaul Thanks! In your provided pipeline, it works well with `"CompVis/stable-diffusion-v1-4"` and `'runwayml/stable-diffusion-v1-5'` but it does not work when I use `'stabilityai/stable-diffusion-2-1-base'`. Could you check that?

**Comment by sayakpaul** [comment]
Could you provide a reproducible snippet for us? That would be very helpful.

**Comment by g-jing** [comment]
Here it is. I first got the caption and the latent variable:

```
import torch
from transformers import BlipForConditionalGeneration, BlipProcessor
from diffusers import DDIMScheduler, DDIMInverseScheduler, StableDiffusionPix2PixZeroPipeline

import requests
from PIL import Image

captioner_id = "Salesforce/blip-image-captioning-base"
processor = BlipProcessor.from_pretrained(captioner_id)
model = BlipForConditionalGeneration.from_pretrained(
    captioner_id, low_cpu_mem_usage=True
)

sd_model_ckpt = 'runwayml/stable-diffusion-v1-5'

pipeline = StableDiffusionPix2PixZeroPipeline.from_pretrained(
    sd_model_ckpt,
    caption_generator=model,
    caption_processor=processor,
    safety_checker=None,
)

pipeline.scheduler = DDIMScheduler.from_config(pipeline.scheduler.config)
pipeline.inverse_scheduler = DDIMInverseScheduler.from_config(pipeline.scheduler.config)
pipeline.enable_model_cpu_offload()

img_url = "https://github.com/pix2pixzero/pix2pix-zero/raw/main/assets/test_images/cats/cat_6.png"

raw_image = Image.open(requests.get(img_url, stream=True).raw).convert("RGB").resize((512, 512))
# generate caption
caption = pipeline.generate_caption(raw_image)

# "a photography of a cat with flowers and dai dai daie - daie - daie kasaii"
inv_latents = pipeline.invert(caption, image=raw_image).latents
# we need to generate source and target embeds

```
Then I use stable diffusion to get the image:
```
import torch
from diffusers import StableDiffusionPipeline

sd_pipe = StableDiffusionPipeline.from_pretrained(sd_model_ckpt, scheduler=pipeline.scheduler)
sd_pipe = sd_pipe.to("cuda")

image = sd_pipe(caption, latents=inv_latents).images[0]
```

**Comment by sayakpaul** [comment]
Got it. Thanks that was very helpful!

For reference, from your example, the original image:

![image](https://user-images.githubusercontent.com/22957388/235091786-7dccac5b-f818-421d-88da-28e4440e00d8.png)


The predicted caption:

> a photography of a black and white kitten in a field of daies

When you call:

```py
import torch
from diffusers import StableDiffusionPipeline

sd_pipe = StableDiffusionPipeline.from_pretrained(sd_model_ckpt, scheduler=pipeline.scheduler)
sd_pipe = sd_pipe.to("cuda")

image = sd_pipe(caption, latents=inv_latents).images[0]
```

I got:

![image](https://user-images.githubusercontent.com/22957388/235091962-d7079d89-063c-43ec-b1fa-444f2e5666d7.png)

I am not sure SD could be conditioned on such inversely predicted noise vectors, though. Cc: @patrickvonplaten @patil-suraj

**Comment by patrickvonplaten** [comment]
Not sure either

**Comment by pikeyang** [comment]
Do you have any new findings?

**Comment by shaibagon** [comment]
A simple example using `diffusers.DDIMInverseScheduler` can be found [here](https://github.com/shaibagon/diffusers_ddim_inversion).

**Comment by yuanzhi-zhu** [comment]
```python
def ddim_inverse_step(scheduler, x, e_t, t):
    next_t = t + scheduler.config.num_train_timesteps // scheduler.num_inference_steps
    a_t = scheduler.alphas_cumprod[t]
    a_next = scheduler.alphas_cumprod[next_t] if next_t <= 999 else torch.tensor(0.)
    sqrt_one_minus_at = torch.sqrt(1 - a_t)
    # current prediction for x_0
    pred_x0 = (x - sqrt_one_minus_at * e_t) / a_t.sqrt()
    # direction pointing to x_t
    dir_xt = (1. - a_next).sqrt() * e_t
    # Equation 12. reversed
    x_next = a_next.sqrt() * pred_x0 + dir_xt
    return x_next, pred_x0

def ddim_inverse_sampling(unet, scheduler, x_0, eta, device):
    images = x_0
    for t in tqdm(scheduler.timesteps):
        # t = 999 - t + 1
        with torch.no_grad():
            model_output = unet(images, t).sample
        images, pred_x0 = ddim_inverse_step(scheduler, images, model_output, t)
    return images
   
ddim_inverse_scheduler = DDIMInverseScheduler(num_train_timesteps=1000, clip_sample=False, steps_offset=1)
ddim_inverse_scheduler.set_timesteps(500)
```
piece of codes that might be helpful
